### ISSUE TITLE
reverse task

### ISSUE THREAD

**Issue by vinnitu** [cosinekitty/geocalc#1]
Hi! 

Very good [article](https://javascript.plainenglish.io/calculating-azimuth-distance-and-altitude-from-a-pair-of-gps-locations-36b4325d8ab0)! 

If we have two (or more) points A.
And we know GPS coordinates every of this points.
Also we have azimuth and altitude to point B from each point A.
I am interesting how to calculate GPS coordinate of point B?

**Comment by cosinekitty** [comment]
Hello @vinnitu , 
I do have a demonstration of the reverse calculation on [this page](http://cosinekitty.com/latlon/). The interesting part of the logic is in [latlon.js](http://cosinekitty.com/latlon/latlon.js). I hope this helps.

**Comment by vinnitu** [comment]
Thanks @cosinekitty for quick reply!

It is not the same because we don't have distance to target. 
And if I understand example right it calculates position on the ground (regardless altitude)

**Comment by cosinekitty** [comment]
Oh... I understand now. I had to read your question again. Let me think about that. It is interesting! This is more like a triangulation problem.

**Comment by cosinekitty** [comment]
@vinnitu I have an idea for an algorithm. I have worked out the math and now I am working on a sample program. It will be a Python program with the following command line parameters:

```
USAGE:  triangulate.py  lat1 lon1 elv1 az1 alt1  lat2 lon2 elv2 az2 alt2

Calculate the best-fit location of a point as observed
from two different locations on or near the Earth's surface.

lat1, lat2 = Geographic latitudes in degrees north of the equator.
lon1, lon2 = Geographic longitudes in degrees east of the prime meridian.
elv1, elv2 = Elevations above sea level in meters.
az1,  az2  = Azimuths toward observed object in degrees clockwise from north.
alt1, alt2 = Altitude angles toward observed object in degrees above horizon.

This program extrapolates lines in the given directions from the two
geographic locations and finds the location in space where they
come closest to intersecting. It then prints out the coordinates
of that triangulation point, along with the error radius in meters.
```

So now I am wondering, do you have some test data I can start with? It would help to have an idea of what your real data looks like.

**Comment by vinnitu** [comment]
something like this

lat,lon,elv,az,alt
48.16042,24.49986,2019,18,7
48.27305,24.36401,662,83,12

target near
~ 48.28821,24.56373,~4000

**Comment by cosinekitty** [comment]
@vinnitu  I have a prototype of this triangulation solver now. I put it in a branch called `vecobj` in the Astronomy Engine project: [triangulate.py](https://github.com/cosinekitty/astronomy/blob/vecobs/demo/python/triangulate.py).

Here is a sample run:

```
$ ./triangulate.py 48.16042 24.49986 2019 18 7 48.27305 24.36401 662 83 12
Solution #1 = (N48.28914299803458, E24.562600140892187, 3862.0102185777796m)
Solution #2 = (N48.28914299803458, E24.562600140892187, 3862.0102185777796m)
```
For now, this is only in the `vecobj` branch. I want to convert this algorithm to the other supported programming languages (C/C++, C#, and TypeScript/JavaScript) before merging it into the `master` branch.

**Comment by cosinekitty** [comment]
I updated `triangulate.py` to also print the distance between the two lines of sight where they come closest to each other. In general, they will not intersect perfectly at a point.

```
$ more test/triangulate.txt
Solution #1 = (N48.28914299803458, E24.562600140892187, 3862.0102185777796m), err = 48.810 meters
Solution #2 = (N48.28914299803458, E24.562600140892187, 3862.0102185777796m), err = 48.810 meters
```

**Comment by vinnitu** [comment]
Great! Seems it works! 

Tell me please how to interpret err value?
Maybe result must be range? In this case [from: 3862.01, to: 3910.82]

And I have strange question...
How do you think, theoretically can altitude angle have negative value? 

How long do you plan implement nodejs version?

**Comment by cosinekitty** [comment]
Yes, the altitude angle can have a negative value if the target object is below the observer's horizon.

**Comment by cosinekitty** [comment]
Hi @vinnitu .  In general, two direction lines seen by two different observers will not intersect at a perfect point. The error radius is half the distance in meters between the two lines, where they come the closest to each other.  If the lines actually do intersect perfectly, then the error radius is 0.

If you look in the `master` branch, you will see I already have a Node.js version for you.  :)  I am currently working on the C# version.  I will push that soon.

**Comment by cosinekitty** [comment]
Here is the Node.js version: [triangulate.js](https://github.com/cosinekitty/astronomy/blob/master/demo/nodejs/triangulate.js)

**Comment by vinnitu** [comment]
What it role of time in this task?

**Comment by cosinekitty** [comment]
That is a good question. The function `ObserverVector` takes a time parameter because it is used for other astronomy functions where I have to include the Earth's rotation. For example, to know when sunset happens for a given observer, I have to know the Earth's rotation with respect to the Sun. In the triangulation problem, the time does not matter, so long as it is consistent, because all the calculations are done inside the Earth's rotating frame of reference. No matter what time value you use, the triangulator should always calculate the same answer.

**Comment by vinnitu** [comment]
excuse me for my insolence ... but I [tried to simplify](https://gist.github.com/vinnitu/af25ed75abfef35349d742cc6f6d4614) that part of your program that deals specifically with my problem. And it seems to me that I have not reached perfection. Can you tell me how this can still be simplified? I think there is something easier to do with terra() and inverse_terra() functions

**Comment by cosinekitty** [comment]
It looks like a great idea for what you are doing. The code is so much simpler. There are a few things you can remove. You do not need the function `InverseRefraction`. For example, you can do:

```JavaScript
function VectorFromHorizon(sphere) {
    const lon = ToggleAzimuthDirection(sphere.lon);
    const lat = sphere.lat;
    const xsphere = new Spherical(lat, lon, sphere.dist);
    return VectorFromSphere(xsphere);
}
```

Most of the complexity is dealing with the non-spherical flattening of the Earth's sea level (the *geoid*). The geodetic latitude we commonly use is not the same as geocentric latitude you would get on a sphere. They are the same at the poles and the equator, but are most different at 45 degrees from the equator (north and south). According to page 2 in [this reference](https://www.titulosnauticos.net/astro/Chapter9.pdf), the maximum error is 11.5 arcminutes. This was far too large for my precision goals for Astronomy Engine.

So here is something you can try that would greatly simplify things, but with the risk of much less accuracy. You will have to test to figure out if it is still accurate enough for your needs. You can replace this:

```JavaScript
const EARTH_FLATTENING = 0.996647180302104;
```

with this:

```JavaScript
const EARTH_FLATTENING = 1;
```

Then go through the code and eliminate everything that knows the Earth is not a sphere. Then `inverse_terra` would become very simple.  I have not tested this, but I think it would look something like:

```JavaScript
function inverse_terra(ovec) {
    const x = ovec[0] * exports.KM_PER_AU;
    const y = ovec[1] * exports.KM_PER_AU;
    const z = ovec[2] * exports.KM_PER_AU;
    const p = Math.sqrt(x * x + y * y);
    let lon_deg, lat_deg, height_km;
    if (p < 1.0e-6) {
        lon_deg = 0;
        lat_deg = (z > 0.0) ? +90 : -90;
    }
    else {
        lon_deg = exports.RAD2DEG * Math.atan2(y, x);
        while (lon_deg <= -180)
            lon_deg += 360;
        while (lon_deg > +180)
            lon_deg -= 360;
        lat_deg = exports.RAD2DEG * Math.atan2(z, p);
    }
    height_km = z - EARTH_EQUATORIAL_RADIUS_KM;
    return new Observer(lat_deg, lon_deg, 1000 * height_km);
}
```
